enhancements/network/ovs-dynamic-cpu-pinning.md

```diff
@@ -0,0 +1,204 @@
+  the feature (see [Activation file](#activation-file))
+
+```mermaid
+sequenceDiagram
+    Note right of NodeTuningOperator: via MachineConfig
+  
+    NodeTuningOperator->>ovs-vswitchd: set_slice
+    NodeTuningOperator->>ovsdb-server: set_slice
+    NodeTuningOperator->>/etc/openvswitch/enable_dynamic_cpu_affinity: fwrite('1')
+    activate /etc/openvswitch/enable_dynamic_cpu_affinity
+    ovnkube-node ->> /etc/openvswitch/enable_dynamic_cpu_affinity: fstat()
+    
+    activate ovnkube-node
+    loop
+    ovnkube-node ->> ovnkube-node: x = sched_getaffinity(os.getPid())
+    ovnkube-node ->> ovs-vswitchd: sched_setaffinity(x)
+    ovnkube-node ->> ovsdb-server: sched_setaffinity(x)
+    ovnkube-node ->> ovsdb-server: sleep(10s)
+    end
+
+    deactivate ovnkube-node
+    deactivate /etc/openvswitch/enable_dynamic_cpu_affinity
+```
+
+### API Extensions
+
+N/A
+
+### Implementation Details/Notes/Constraints
+
+#### CPU affinity alignment
+
+CPU affinity alignment is implemented using system calls SYS_SCHED_SETAFFINITY and SYS_SCHED_GETAFFINITY.
+Pods from `ovnkube-node` daemonset run in the Burstable Quality of Service (QoS) class, hence their CPU 
+affinities are managed by the [kubelet cpu manager](https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/) and it has access to all cpus that are not being used by guaranteed and pinned pods.
+
+Note that the affinity alignement might take a few second to happen as the CPU reconciler will take few 
+seconds to update the affinity of the ovnkube-node container and then another few seconds until ovnkube-node
+ otices and updates the affinity of OVS daemons. During this window a newly started guaranteed and cpu pinned container might encounter latency spikes.
+
+#### Activation file
+
+The file `/etc/openvswitch/enable_dynamic_cpu_affinity` on the host filesystem is used to 
+activate the feature on ovnkube-node according to the following logic:
+a. If the file is missing OR the file file is empty, then the feature is turned `OFF`.
+b. If the file exists and its content is not an empty string, then the feature is turned `ON`.
+c. `ovnkube-node` responds to the file modifications (deletion or content update)
+  at runtime, without the need for a restart.
+
+#### Must-Gather enhancement
+
+As the feature is enabled via a file on the host filesystem, it's useful to
+add `/etc/openvswitch/enable_dynamic_cpu_affinity` file contents to the [gather_network_logs](https://github.com/openshift/must-gather/blob/master/collection-scripts/gather_network_logs) script.
+
+### Risks and Mitigations
+
+|Risk|Mitigation|
+|---|---|
+|After NTO enabled the feature, a network performance decrease occurred in some cluster nodes.|The cluster administrator can turn off the feature on specific nodes by connecting through SSH and deleting the activation file. An overriding MachineConfig can be created to achieve the same goal.|
```

Comment:
I was wondering if we need to suggest deleting the file or rather deleting its contents. I think either will do. I've just done a few tests on a live cluster, since I didn't remember how MCO reacts to file changes. It turns out... MCO realizes a file from a MachineConfig is missing or has different contents than what it's supposed to have, but then it doesn't take any action to try to reconcile this file. So this plays in our favour. However, the result is that MCO will show a `degraded` status in this case. Is this OK? I suppose it is, since the action we're describing here is a debug/troubleshooting action, but it's better to know this beforehand :)

Reply:
This MCO behaviour is documented here https://docs.google.com/document/d/1fgP6Kv1D-75e1Ot0Kg-W2qPyxWDp2_CALltlBLuseec/edit#heading=h.phy2b7q0mu4

Reply:
Yep, the SSH approach will cause this. The override MCO cannot delete a file, but it can change the contents (that is why an empty file disables the feature as well).